cmd/sfncli/cloudwatchreporter.go

```diff
@@ -11,55 +11,105 @@ import (
 	"gopkg.in/Clever/kayvee-go.v6/logger"
 )
 
+const metricNameActivityActivePercent = "ActivityActivePercent"
+const namespaceStatesCustom = "StatesCustom"
+
 // CloudWatchReporter reports useful metrics about the activity.
 type CloudWatchReporter struct {
 	cwapi       cloudwatchiface.CloudWatchAPI
 	activityArn string
-	idleTime    int64
+
+	// state to keep track of active percent
+	mu                    sync.Mutex
+	activeState           bool
+	activeTime            time.Duration
+	lastReportingTime     time.Time
+	lastActiveStateChange time.Time
 }
 
 func NewCloudWatchReporter(cwapi cloudwatchiface.CloudWatchAPI, activityArn string) *CloudWatchReporter {
+	now := time.Now()
 	c := &CloudWatchReporter{
 		cwapi:       cwapi,
 		activityArn: activityArn,
+
+		activeState:           false,
+		activeTime:            time.Duration(0),
+		lastReportingTime:     now,
+		lastActiveStateChange: now,
 	}
 	return c
 }
 
-// ReportIdleTime reports time spent idle to cloudwatch as a counter.
+// ReportActivePercent sets up a loop that will report active percent to cloudwatch on an interval.
 // It stops when the context is canceled.
-func (cwr *CloudWatchReporter) ReportIdleTime(ctx context.Context, interval time.Duration) {
+func (c *CloudWatchReporter) ReportActivePercent(ctx context.Context, interval time.Duration) {
 	ticker := time.NewTicker(interval)
 	defer ticker.Stop()
 	for ctx.Err() == nil {
 		select {
 		case <-ctx.Done():
 			break
 		case <-ticker.C:
-			val := atomic.SwapInt64(&cwr.idleTime, 0) // reset counter to zero
-			log.InfoD("reportidletime", logger.M{"seconds": time.Duration(val) / time.Second})
-			if _, err := cwr.cwapi.PutMetricDataWithContext(ctx, &cloudwatch.PutMetricDataInput{
-				MetricData: []*cloudwatch.MetricDatum{{
-					Dimensions: []*cloudwatch.Dimension{{
-						Name:  aws.String("ActivityArn"),
-						Value: aws.String(cwr.activityArn),
-					}},
-					MetricName: aws.String("ActivityIdleTimeSeconds"),
-					Unit:       aws.String(cloudwatch.StandardUnitSeconds),
-					Value:      aws.Float64(float64(time.Duration(val) / time.Second)),
-				}},
-				Namespace: aws.String("StatesCustom"),
-			}); err != nil && err != context.Canceled {
-				log.ErrorD("reportidletime-error", logger.M{"error": err.Error()})
-				// add back value to account for failure to record it in cloudwatch
-				atomic.AddInt64(&cwr.idleTime, val)
-			}
+			c.report()
 		}
 	}
-	log.Info("reportidletime-stop")
 }
 
-// CountIdleTime records time spent idle.
-func (cwr *CloudWatchReporter) CountIdleTime(idle time.Duration) {
-	atomic.AddInt64(&cwr.idleTime, int64(idle))
+// SetActiveState sets whether the activity is currently working on a task or not.
+func (c *CloudWatchReporter) SetActiveState(active bool) {
+	c.mu.Lock()
```

Comment:
would be nice to describe why a mutex is useful here.

Reply:
described the reason 0ca1b4e

Reply:
although on second thought the comment in that commit seems obvious. Can you be more specific about what kind of description you're looking for?

Reply:
Yeah usually I think comments about what is the contention or the multiple go-routines that necessitate  a mutex is helpful.

Example: _These fields are written to by go routines when reporting and when completing a job._